File: golang/internal/engine/rules.go
package engine

import (
	"context"
	"fmt"
	"regexp"

	"github.com/expr-lang/expr"
	"github.com/expr-lang/expr/vm"

	"rocket-backend/internal/instrument"
	"rocket-backend/internal/metadata"
)

// EvaluateRules runs all active rules for an entity/hook against the record.
// It returns validation errors for field and expression rules, and mutates
// the fields map for computed rules.
func EvaluateRules(ctx context.Context, reg *metadata.Registry, entityName string, hook string, fields map[string]any, old map[string]any, isCreate bool) []ErrorDetail {
	_, span := instrument.GetInstrumenter(ctx).StartSpan(ctx, "engine", "rules", "rules.evaluate")
	defer span.End()
	span.SetEntity(entityName, "")

	rules := reg.GetRulesForEntity(entityName, hook)
	if len(rules) == 0 {
		span.SetStatus("ok")
		return nil
	}

	action := "update"
	if isCreate {
		action = "create"
	}

	env := map[string]any{
		"record": fields,
		"old":    old,
		"action": action,
	}

	var errs []ErrorDetail

	// 1. Field rules
	for _, r := range rules {
		if r.Type != "field" {
			continue
		}
		if detail := EvaluateFieldRule(r, fields); detail != nil {
			errs = append(errs, *detail)
			if r.Definition.StopOnFail {
				span.SetStatus("error")
				return errs
			}
		}
	}

	// 2. Expression rules
	for _, r := range rules {
		if r.Type != "expression" {
			continue
		}
		if detail := EvaluateExpressionRule(r, env); detail != nil {
			errs = append(errs, *detail)
			if r.Definition.StopOnFail {
				span.SetStatus("error")
				return errs
			}
		}
	}

	// If there are validation errors, don't run computed fields
	if len(errs) > 0 {
		span.SetStatus("error")
		return errs
	}

	// 3. Computed fields
	for _, r := range rules {
		if r.Type != "computed" {
			continue
		}
		val, err := EvaluateComputedField(r, env)
		if err != nil {
			errs = append(errs, ErrorDetail{
				Field:   r.Definition.Field,
				Rule:    "computed",
				Message: err.Error(),
			})
			continue
		}
		fields[r.Definition.Field] = val
	}

	if len(errs) > 0 {
		span.SetStatus("error")
	} else {
		span.SetStatus("ok")
	}
	return errs
}

// EvaluateFieldRule evaluates a single field rule against a record.
// Returns nil if the rule passes, or an ErrorDetail if it fails.
func EvaluateFieldRule(rule *metadata.Rule, record map[string]any) *ErrorDetail {
	fieldName := rule.Definition.Field
	val, exists := record[fieldName]
	if !exists || val == nil {
		return nil // absent fields are not checked by field rules (use "required" for that)
	}

	op := rule.Definition.Operator
	msg := rule.Definition.Message
	if msg == "" {
		msg = fmt.Sprintf("field %s failed %s validation", fieldName, op)
	}

	switch op {
	case "min":
		num, ok := toFloat64(val)
		if !ok {
			return nil
		}
		threshold, ok := toFloat64(rule.Definition.Value)
		if !ok {
			return nil
		}
		if num < threshold {
			return &ErrorDetail{Field: fieldName, Rule: "min", Message: msg}
		}

	case "max":
		num, ok := toFloat64(val)
		if !ok {
			return nil
		}
		threshold, ok := toFloat64(rule.Definition.Value)
		if !ok {
			return nil
		}
		if num > threshold {
			return &ErrorDetail{Field: fieldName, Rule: "max", Message: msg}
		}

	case "min_length":
		s, ok := val.(string)
		if !ok {
			return nil
		}
		threshold, ok := toFloat64(rule.Definition.Value)
		if !ok {
			return nil
		}
		if len(s) < int(threshold) {
			return &ErrorDetail{Field: fieldName, Rule: "min_length", Message: msg}
		}

	case "max_length":
		s, ok := val.(string)
		if !ok {
			return nil
		}
		threshold, ok := toFloat64(rule.Definition.Value)
		if !ok {
			return nil
		}
		if len(s) > int(threshold) {
			return &ErrorDetail{Field: fieldName, Rule: "max_length", Message: msg}
		}

	case "pattern":
		s, ok := val.(string)
		if !ok {
			return nil
		}
		pattern, ok := rule.Definition.Value.(string)
		if !ok {
			return nil
		}
		matched, err := regexp.MatchString(pattern, s)
		if err != nil || !matched {
			return &ErrorDetail{Field: fieldName, Rule: "pattern", Message: msg}
		}
	}

	return nil
}

// CompileExpression compiles an expression string into an expr-lang program.
func CompileExpression(expression string) (*vm.Program, error) {
	prog, err := expr.Compile(expression, expr.AsBool())
	if err != nil {
		return nil, fmt.Errorf("compile expression: %w", err)
	}
	return prog, nil
}

// EvaluateExpressionRule evaluates a compiled expression rule against an environment.
// The env should contain: record, old, action (and optionally related data).
// Returns nil if the rule passes (expression is false), or an ErrorDetail if violated (expression is true).
func EvaluateExpressionRule(rule *metadata.Rule, env map[string]any) *ErrorDetail {
	prog, ok := rule.Compiled.(*vm.Program)
	if !ok || prog == nil {
		// Lazy compile
		compiled, err := CompileExpression(rule.Definition.Expression)
		if err != nil {
			return &ErrorDetail{Rule: "expression", Message: fmt.Sprintf("compile error: %v", err)}
		}
		rule.Compiled = compiled
		prog = compiled
	}

	result, err := expr.Run(prog, env)
	if err != nil {
		return &ErrorDetail{Rule: "expression", Message: fmt.Sprintf("rule evaluation error: %v", err)}
	}

	violated, ok := result.(bool)
	if !ok {
		return nil
	}

	if violated {
		msg := rule.Definition.Message
		if msg == "" {
			msg = "Expression rule violated"
		}
		return &ErrorDetail{Rule: "expression", Message: msg}
	}

	return nil
}

// CompileComputedExpression compiles an expression for a computed field (returns any value, not bool).
func CompileComputedExpression(expression string) (*vm.Program, error) {
	prog, err := expr.Compile(expression)
	if err != nil {
		return nil, fmt.Errorf("compile computed expression: %w", err)
	}
	return prog, nil
}

// EvaluateComputedField evaluates a computed field rule and returns the computed value.
func EvaluateComputedField(rule *metadata.Rule, env map[string]any) (any, error) {
	prog, ok := rule.Compiled.(*vm.Program)
	if !ok || prog == nil {
		// Lazy compile
		compiled, err := CompileComputedExpression(rule.Definition.Expression)
		if err != nil {
			return nil, fmt.Errorf("compile computed expression: %w", err)
		}
		rule.Compiled = compiled
		prog = compiled
	}

	result, err := expr.Run(prog, env)
	if err != nil {
		return nil, fmt.Errorf("evaluate computed field %s: %w", rule.Definition.Field, err)
	}

	return result, nil
}

// toFloat64 converts numeric types to float64.
func toFloat64(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case int32:
		return float64(n), true
	}
	return 0, false
}

File: backend/expressjs/src/engine/rules.ts
import type { Rule } from "../metadata/rule.js";
import type { Registry } from "../metadata/registry.js";
import type { ErrorDetail } from "./errors.js";
import { getInstrumenter } from "../instrument/instrument.js";

/**
 * Evaluates all active rules for an entity/hook against the record.
 * Returns validation errors for field and expression rules.
 * Mutates the fields map for computed rules.
 */
export function evaluateRules(
  registry: Registry,
  entityName: string,
  hook: string,
  fields: Record<string, any>,
  old: Record<string, any>,
  isCreate: boolean,
): ErrorDetail[] {
  const span = getInstrumenter().startSpan("engine", "rules", "rules.evaluate");
  span.setEntity(entityName);
  span.setMetadata("hook", hook);
  try {
    const rules = registry.getRulesForEntity(entityName, hook);
    if (rules.length === 0) {
      span.setStatus("ok");
      span.setMetadata("rule_count", 0);
      return [];
    }

    const action = isCreate ? "create" : "update";
    const env = { record: fields, old, action };
    const errs: ErrorDetail[] = [];

    // 1. Field rules
    for (const r of rules) {
      if (r.type !== "field") continue;
      const detail = evaluateFieldRule(r, fields);
      if (detail) {
        errs.push(detail);
        if (r.definition.stop_on_fail) {
          span.setStatus("ok");
          span.setMetadata("rule_count", rules.length);
          span.setMetadata("error_count", errs.length);
          return errs;
        }
      }
    }

    // 2. Expression rules
    for (const r of rules) {
      if (r.type !== "expression") continue;
      const detail = evaluateExpressionRule(r, env);
      if (detail) {
        errs.push(detail);
        if (r.definition.stop_on_fail) {
          span.setStatus("ok");
          span.setMetadata("rule_count", rules.length);
          span.setMetadata("error_count", errs.length);
          return errs;
        }
      }
    }

    // If there are validation errors, don't run computed fields
    if (errs.length > 0) {
      span.setStatus("ok");
      span.setMetadata("rule_count", rules.length);
      span.setMetadata("error_count", errs.length);
      return errs;
    }

    // 3. Computed fields
    for (const r of rules) {
      if (r.type !== "computed") continue;
      try {
        const val = evaluateComputedField(r, env);
        fields[r.definition.field!] = val;
      } catch (err: any) {
        errs.push({
          field: r.definition.field,
          rule: "computed",
          message: err.message ?? String(err),
        });
      }
    }

    span.setStatus("ok");
    span.setMetadata("rule_count", rules.length);
    span.setMetadata("error_count", errs.length);
    return errs;
  } catch (err) {
    span.setStatus("error");
    span.setMetadata("error", (err as Error).message);
    throw err;
  } finally {
    span.end();
  }
}

/**
 * Evaluates a single field rule against a record.
 */
export function evaluateFieldRule(
  rule: Rule,
  record: Record<string, any>,
): ErrorDetail | null {
  const fieldName = rule.definition.field!;
  const val = record[fieldName];
  if (val === undefined || val === null) return null;

  const op = rule.definition.operator!;
  const msg =
    rule.definition.message ??
    `field ${fieldName} failed ${op} validation`;

  switch (op) {
    case "min": {
      const num = toNumber(val);
      const threshold = toNumber(rule.definition.value);
      if (num === null || threshold === null) return null;
      if (num < threshold) {
        return { field: fieldName, rule: "min", message: msg };
      }
      break;
    }
    case "max": {
      const num = toNumber(val);
      const threshold = toNumber(rule.definition.value);
      if (num === null || threshold === null) return null;
      if (num > threshold) {
        return { field: fieldName, rule: "max", message: msg };
      }
      break;
    }
    case "min_length": {
      if (typeof val !== "string") return null;
      const threshold = toNumber(rule.definition.value);
      if (threshold === null) return null;
      if (val.length < threshold) {
        return { field: fieldName, rule: "min_length", message: msg };
      }
      break;
    }
    case "max_length": {
      if (typeof val !== "string") return null;
      const threshold = toNumber(rule.definition.value);
      if (threshold === null) return null;
      if (val.length > threshold) {
        return { field: fieldName, rule: "max_length", message: msg };
      }
      break;
    }
    case "pattern": {
      if (typeof val !== "string") return null;
      const pattern = rule.definition.value;
      if (typeof pattern !== "string") return null;
      try {
        const re = new RegExp(pattern);
        if (!re.test(val)) {
          return { field: fieldName, rule: "pattern", message: msg };
        }
      } catch {
        return { field: fieldName, rule: "pattern", message: msg };
      }
      break;
    }
  }
  return null;
}

/**
 * Evaluates an expression rule against an environment.
 * Expression returning true = rule violated.
 */
export function evaluateExpressionRule(
  rule: Rule,
  env: Record<string, any>,
): ErrorDetail | null {
  const expression = rule.definition.expression;
  if (!expression) return null;

  try {
    // Simple expression evaluator using Function constructor
    // Environment variables are passed as named params
    const fn = compileExpression(rule, expression);
    const result = fn(env);

    if (result === true) {
      const msg = rule.definition.message ?? "Expression rule violated";
      return { rule: "expression", message: msg };
    }
  } catch (err: any) {
    return {
      rule: "expression",
      message: `rule evaluation error: ${err.message ?? err}`,
    };
  }
  return null;
}

/**
 * Evaluates a computed field expression and returns the computed value.
 */
export function evaluateComputedField(
  rule: Rule,
  env: Record<string, any>,
): any {
  const expression = rule.definition.expression;
  if (!expression) throw new Error("computed rule has no expression");

  const fn = compileComputedExpression(rule, expression);
  return fn(env);
}

// --- Expression compilation with caching ---

function compileExpression(
  rule: Rule,
  expression: string,
): (env: Record<string, any>) => boolean {
  if (rule.compiled) return rule.compiled as (env: Record<string, any>) => boolean;

  const fn = new Function(
    "env",
    `with (env) { return !!(${expression}); }`,
  ) as (env: Record<string, any>) => boolean;
  rule.compiled = fn;
  return fn;
}

function compileComputedExpression(
  rule: Rule,
  expression: string,
): (env: Record<string, any>) => any {
  if (rule.compiled) return rule.compiled as (env: Record<string, any>) => any;

  const fn = new Function(
    "env",
    `with (env) { return (${expression}); }`,
  ) as (env: Record<string, any>) => any;
  rule.compiled = fn;
  return fn;
}

function toNumber(v: any): number | null {
  if (typeof v === "number") return v;
  if (typeof v === "string") {
    const n = Number(v);
    return isNaN(n) ? null : n;
  }
  return null;
}
